### ISSUE TITLE
export to 2 modules from 1 notebook

### ISSUE THREAD

**Issue by raynardj** [fastai/nbdev#79]
* I was trying to export 2 modules from 1 single notebook, as I learnt from the doc [here](http://nbdev.fast.ai/export/index.html#is_export) 
* Upon ```nbdev_build_lib``` , I ran into error
```
Traceback (most recent call last):
  File "/Users/salvor/anaconda3/bin/nbdev_build_lib", line 8, in <module>
    sys.exit(nbdev_build_lib())
  File "/Users/salvor/.local/lib/python3.7/site-packages/fastscript/core.py", line 73, in _f
    func(**args.__dict__)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/cli.py", line 22, in nbdev_build_lib
    notebook2script(fname=fname)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 371, in notebook2script
    for f in sorted(files): d = _notebook2script(f, silent=silent, to_dict=d)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 310, in _notebook2script
    if to_dict is None: _add2add(fname_out, [f"'{f}'" for f in names if '.' not in f and len(f) > 0] + extra)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 183, in _add2add
    start,end = re_all.start(),re_all.end()
AttributeError: 'NoneType' object has no attribute 'start'
```
* I realize ```#default_exp [module]``` can solve this, as this call the function ```create_mod_file``` to create the ```__all__``` expression , but put 2 ```#default_exp``` cells in 1 notebook will make the 2nd one  ignored.

* I also tried to avoid this by creating a notebook with a single cell specifying  a second ```#default_exp [module]```, I can run ```nbdev_build_lib``` smoothly, but  ```nbdev_build_docs``` run into further error
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-2-e7b5de274890> in <module>
----> 1 show_doc(docModel, default_cls_level=2)
                                                                                      
NameError: name 'docModel' is not defined

  File "/Users/salvor/anaconda3/bin/nbdev_build_docs", line 8, in <module>
    sys.exit(nbdev_build_docs())
  File "/Users/salvor/.local/lib/python3.7/site-packages/fastscript/core.py", line 73, in _f
    func(**args.__dict__)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/cli.py", line 164, in nbdev_build_docs
    notebook2html(fname=fname, force_all=force_all, n_workers=n_workers)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export2html.py", line 412, in notebook2html
    raise Exception(msg + '\n'.join([f.name for p,f in zip(passed,files) if not p]))
Exception: Conversion failed on the following:
python_parse.ipynb
```

Please pardon my stubborn about this, I just find generate 2 modules from 1 notebook useful, I can expand an entire project from 1 experimenting  notebook easily. Thanks

**Comment by sgugger** [comment]
The library doesn't support having two different `default_exp` values. To export a cell in a different module than the default you set, set the flag `# export module_name`. If you want to export one given to cell to two different modules, you will need to duplicate it with two different export flags for now.
Hope this helps solve your issue!

**Comment by raynardj** [comment]
> `ult_exp` values. To export a cell in a different module than the default you set, set the flag `# export module_name`. If you want to export one given to cell to two different modules, you will need to duplicate it with two different export flags for now.
> Hope this helps solve your issue!

Thanks for the reply,  yes I realize there is ```# export module_name ```
But this will run in to error when ```nbdev_build_lib```
```
AttributeError: 'NoneType' object has no attribute 'start'
```

**Comment by sgugger** [comment]
What is the exact comment you type, so that I have a reproducer?

**Comment by raynardj** [comment]
### Reproduce the ```No such file or directory``` error
eg. the [dummy notebook](https://github.com/raynardj/forgebox/blob/master/nbs/test_nbdev.ipynb) I was trying to convert,

I ran ```nbdev_build_lib```, the following is the error:
```
Traceback (most recent call last):
  File "/Users/salvor/anaconda3/bin/nbdev_build_lib", line 8, in <module>
    sys.exit(nbdev_build_lib())
  File "/Users/salvor/.local/lib/python3.7/site-packages/fastscript/core.py", line 73, in _f
    func(**args.__dict__)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/cli.py", line 22, in nbdev_build_lib
    notebook2script(fname=fname)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 371, in notebook2script
    for f in sorted(files): d = _notebook2script(f, silent=silent, to_dict=d)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 310, in _notebook2script
    if to_dict is None: _add2add(fname_out, [f"'{f}'" for f in names if '.' not in f and len(f) > 0] + extra)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 180, in _add2add
    with open(fname, 'r', encoding='utf8') as f: text = f.read()
FileNotFoundError: [Errno 2] No such file or directory: '/Users/salvor/github/forgebox/forgebox/different_file.py
```

However, the [test_nbdev.py](https://github.com/raynardj/forgebox/blob/master/forgebox/test_nbdev.py) was built successfully.

### Try the empty file
I guess the pre-existence is important for ```different_file.py```, so I created such empty file, run ```nbdev_build_lib```,  following the new error
```
Traceback (most recent call last):
  File "/Users/salvor/anaconda3/bin/nbdev_build_lib", line 8, in <module>
    sys.exit(nbdev_build_lib())
  File "/Users/salvor/.local/lib/python3.7/site-packages/fastscript/core.py", line 73, in _f
    func(**args.__dict__)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/cli.py", line 22, in nbdev_build_lib
    notebook2script(fname=fname)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 371, in notebook2script
    for f in sorted(files): d = _notebook2script(f, silent=silent, to_dict=d)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 310, in _notebook2script
    if to_dict is None: _add2add(fname_out, [f"'{f}'" for f in names if '.' not in f and len(f) > 0] + extra)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export.py", line 183, in _add2add
    start,end = re_all.start(),re_all.end()
AttributeError: 'NoneType' object has no attribute 'start'
```
### When using another notebook's ```default_exp```
Okay, so the empty one won't suffice, so I created [another notebook](https://github.com/raynardj/forgebox/blob/master/nbs/create_different.ipynb) , solely in purpose to create a fully fledged py file.

This time, ```nbdev_build_lib``` turns out fine. all py file are in place

But ```nbdev_build_docs``` will have this error
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-e86c8b69e1ff> in <module>
----> 1 show_doc(different_file_func, default_cls_level=2)

NameError: name 'different_file_func' is not defined
```

**Comment by sgugger** [comment]
You have to put a number of your notebooks since they need to run in a certain order during nbdev_build_lib:
- the notebook creating different_file.py has to run first so that the module exists
- then the notebook creating test_file.py can run and add its function different_file_func to different_file

This is why all notebooks in our nbdev-generated packages begin with digits.

**Comment by raynardj** [comment]
yes I tried,  the correct order can generate py files correctly.  yes

but ** nbdev_build_docs** will have error ```NameError: name 'different_file_func' is not defined```

**Comment by sgugger** [comment]
I need more than this as I can't reproduce what you are saying. Please share a whole folder with the numbered notebooks, exported library where the error appears.

**Comment by raynardj** [comment]
I added the following 2 files into a [nbdev template repo](https://github.com/raynardj/nbdev_test)

* [01_create_different.ipynb](https://github.com/raynardj/nbdev_test/blob/master/01_create_different.ipynb), this will create a file using ```#default_exp different_file```
* [02test_nbdev.ipynb](02test_nbdev.ipynb), inserting a function to ```different_file.py```

#### Then in this step, everything's okay 👌
```$ nbdev_build_lib```
#### Then in this step for docs, error ❌
```$ nbdev_build_docs```

```
converting: /Users/salvor/github/nbdev_test/00_core.ipynb
converting: /Users/salvor/github/nbdev_test/02test_nbdev.ipynb
converting: /Users/salvor/github/nbdev_test/index.ipynb
converting: /Users/salvor/github/nbdev_test/01_create_different.ipynb
An error occurred while executing the following cell:
------------------
show_doc(different_file_func, default_cls_level=2)
------------------

---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-3-e86c8b69e1ff> in <module>
----> 1 show_doc(different_file_func, default_cls_level=2)

NameError: name 'different_file_func' is not defined
NameError: name 'different_file_func' is not defined

Traceback (most recent call last):
  File "/Users/salvor/anaconda3/bin/nbdev_build_docs", line 8, in <module>
    sys.exit(nbdev_build_docs())
  File "/Users/salvor/.local/lib/python3.7/site-packages/fastscript/core.py", line 73, in _f
    func(**args.__dict__)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/cli.py", line 164, in nbdev_build_docs
    notebook2html(fname=fname, force_all=force_all, n_workers=n_workers)
  File "/Users/salvor/.local/lib/python3.7/site-packages/nbdev/export2html.py", line 412, in notebook2html
    raise Exception(msg + '\n'.join([f.name for p,f in zip(passed,files) if not p]))
Exception: Conversion failed on the following:
02test_nbdev.ipynb
```

**Comment by CEM3H** [comment]
I encountered the same issue when trying to generate two (or more) .py files from single notebook - I thought it was a good idea to place function definitions and other code into different modules. 
So even when I bypassed `No such file or directory` error (but its behavior is still unclear sometimes)

I noticed that modules created by `# export <module>` expression expand with each run of `nbdev_build_lib` which duplicates code and thus quickly makes these modules become messy if one runs `nbdev_build_lib` or `notebook2script` often (e.g. as a part of iterative process)

So I found that feature for cleaning `lib_path` might be useful as an argument for `notebook2script` to re-create all the modules along with `_ndbev.py`. 

It also occured to me that a little more explicit behavior of creating modules might be useful - I believe it could be achieved by calling `create_mod_file` func inside `_notebook2script` when iterating over `cells` when module in `# export <module>` still wasn't created during conversion.
